### ISSUE TITLE
Question about sqrt(H*W) scaling in FACL implementation

### ISSUE THREAD

**Issue by liname3** [argenycw/FACL#3]
Hi authors, thanks for releasing the code for FACL.

While reading the implementation, I noticed the loss is scaled by a resolution-dependent factor:

_, _, _, H, W = pred.shape
weight = np.sqrt(H * W)
loss = prob * FAL(fft_pred, fft_gt) + (1 - prob) * FCL(fft_pred, fft_gt)
loss = loss * weight

In the paper, FAL is defined with an explicit 1/(MN) normalization over the Fourier plane, and FCL is a normalized correlation (so its scale should already be bounded). I couldn’t find a mention of an additional sqrt(MN) scaling in the paper.

Could you please clarify:

What is the motivation for multiplying the FACL loss by sqrt(H*W)?

Is it for stabilizing gradients / matching the magnitude to another loss / compensating for a specific FFT normalization?

In the released code, FAL is implemented via nn.MSELoss() on abs(fft) which averages over all tensor dimensions by default (including batch/time/channel).

Is this intended, or should the normalization be only over the Fourier plane (H,W) (and then averaged over batch/time/channel separately) to match the paper’s 1/(MN)?

If there is a recommended implementation to match the paper exactly (especially regarding normalization and scaling), I’d really appreciate it. Thanks!

**Comment by sqfoo** [comment]
Hi, @liname3 

Thank you for appreciating our work.

In our initial experiments, we tested out **FACL** with the default **backward** normalisation (which means no normalisation, suggested by [torch.fft.fftn](https://docs.pytorch.org/docs/stable/generated/torch.fft.fftn.html). While, in [the released code](https://github.com/argenycw/FACL/blob/main/utilspp.py#L603-L604 ), we switched to orthonormal normalisation by setting ```norm='ortho'```. So, to keep the results consistent with our initial setup, we multiply $$\sqrt{H\cdot W}$$ as the compensation factor.

Second, the formulation of **FAL** in the paper exactly corresponds to ```nn.MSELoss()``` applied to ```abs(fft)``` in [the released code](https://github.com/argenycw/FACL/blob/main/utilspp.py#L593-L594 ). $$\sqrt{H\cdot W}$$ is not related to **FAL**.

[```RandomScheduling```](https://github.com/argenycw/FACL/blob/main/utilspp.py#L563-L613) is exactly the **FACL** loss function. So, you can simply follow the ```Training``` Section in ```README.md```, which will call ```RandomScheduling``` for implementing **FACL**.

**Comment by liname3** [comment]
Hi,@sqfoo 
 
Thanks again for the quick reply !

Just to update: I was able to successfully reproduce your reported results on the benchmark setting.

Now I’m applying FACL (RandomScheduling) to my own data, where the spatial resolution is much higher (H=1800, W=3600). In this setting, the FACL loss becomes very large (typically > 1200).

What I observe is: training looks great in the early stage, but when the scheduling gradually transitions to the FAL branch, the training starts to collapse: outputs become progressively darker and fine details degrade over time (sometimes it eventually diverges).

My main question is:
1) Is the absolute magnitude/scale of the FACL loss expected to matter for optimization stability, especially during the transition to FAL?
2) In this high-resolution case, do you recommend any specific practices (e.g. adjusting learning rate proportionally, gradient clipping, or normalizing the loss) to avoid this “darkening / detail collapse” behavior when moving toward FAL?

If needed I can share a short training log (loss curve + example frames) or gradient norms around the transition point.

Thanks a lot!

**Comment by sqfoo** [comment]
Hi @liname3 

It is glad to see that you could reproduce the reported results.

We haven't applied **FACL** at very high resolutions (i.e., H=1800, W=3600). The highest spatial resolution that we tested out so far is H=W=480. 

A few tips that might help:

1. Add a sigmoid layer after the model such that the predictions are in [0, 1].
2. Use a smaller $\alpha$ for datasets where the relative high frequency components does not matter much:
    - $\alpha=0.1$ for the SEVIR and HKO-7 dataset
    - $\alpha=0.3$ for the Stochastic Moving dataset.
3. You can also try using the default ```backward``` FFT normalisation with ```weight=1```.

If possible, please share a short training log. That would help to tune the hyper-parameters.

**Comment by liname3** [comment]
Hello again @sqfoo ,

Thank you so much for your continued help — I really appreciate your patience and the detailed suggestions.

Key details / hyper-parameters (following your tips):
- Data scaling: inputs are normalized to [0, 1]
- Output activation: sigmoid (predictions constrained to [0, 1])
- FACL: alpha (ratio) = 0.2
- total_steps = 21780 (10 epochs * 2178 steps/epoch)
- Rollout: multi-step rollout = 3
- Scheduler: CosineAnnealingLR (T_max=21780, eta_min=1e-6，update persteps)

Attachments:
- Curves: (1) FACL train loss , (2) val MAE 
- A few qualitative visualizations (GT vs pred )

Questions:
1) I’m currently using rolling prediction with a fixed 3-step rollout (i.e., each training iteration unrolls 3 steps). For this setting, do you recommend switching between FAL and FCL within the same iteration (across the 3 unrolled steps), or keeping the same loss mode fixed within an iteration and only switching across iterations?
2) Could you please take a quick look at the attached logs and let me know if anything looks off (e.g., loss scale, schedule, phase transition), and what hyper-parameters you would adjust first for this high-resolution setup?

Thanks again for your time and help.

[logs & vis.zip](https://github.com/user-attachments/files/24520661/logs.vis.zip)

**Comment by sqfoo** [comment]
Hi @liname3 !

From the shared training logs, we're seeing a pattern that differs from ours,  i.e., Validation MAE and MSE always degrade at the start of fine-tuning, regardless of the datasets.

There shouldn't be a difference between the two settings. PredRNN (trained with **Schedule Sampling**) also benefits from FACL, which is computed using its predictions and ground truth.

Suggestions:
- Try a smaller $\alpha$ (e.g., 0.1 or 0.05) to shorten the FAL fine-tuning process, or even set $\alpha=0$ to skip this fine-tuning. 
   **FYI**: We've observed that the SEVIR's Fourier energy is concentrated in low frequencies, which reduces the benefit of strong high-frequency emphasis, as illustrated below.
 
     <img width="386" height="396" alt="Image" src="https://github.com/user-attachments/assets/8709cf25-d937-4d49-98bc-ae4b87715c12" />

**Comment by liname3** [comment]
Hi @sqfoo ,

Thank you so much for your thoughtful help and the time you took to look into this — I truly appreciate it.I’m going to close this issue, but please feel free to reopen it  if you think there’s anything else we should adjust or document.

Thanks again for the support! 🙏
